odoo/tools/date_utils.py

```diff
@@ -59,6 +82,120 @@ def to_timezone(tz: tzinfo | None) -> Callable[[datetime], datetime]:
+    - weekdays are handled similarly
+      - "=tuesday" sets to Tuesday of the current week (starting Monday) at midnight
+      - "+monday" goes to next Monday (no change if we are on Monday)
+
+    The DSL for relative dates is as follows:
+    ```
+    relative_date := ('today' | 'now')? offset*
+    offset := date_rel | time_rel | weekday
+    date_rel := (regex) [=+-]\d+[dwmy]
+    time_rel := (regex) [=+-]\d+[HMS]
+    weekday := [=+-] ('monday' | ... | 'sunday')
+    ```
+
+    An equivalent function is JavaScript is `parseSmartDateInput`.
+
+    :param value: The string to parse
+    :param env: The environment to get the current date (in user's tz)
+    :param naive: Whether to cast the result to a naive datetime.
+    """
+    if re.match(r'\d+-', value):
+        return parse_iso_date(value)
+    terms = value.split()
+    if not terms:
+        raise ValueError("Empty date value")
+
+    # Find the starting point
+    from odoo.orm.fields_temporal import Date, Datetime  # noqa: PLC0415
+    now: datetime | date = Datetime.now()
+    if terms[0] == 'today':
+        terms.pop(0)
+        now = Date.context_today(env['base'], now)
+    else:
+        if terms[0] == 'now':
+            terms.pop(0)
+        now = Datetime.context_timestamp(env['base'], now)
+
+    for term in terms:
+        operator = term[0]
+        if operator not in ('+', '-', '=') or len(term) < 3:
+            raise ValueError(f"Invalid term {term!r} in expression date: {value!r}")
+        # Weekday
+        if (weekday_offset := WEEKDAY_NUMBER.get(term[1:])) is not None:
+            # current week starting on Monday
+            weekday_offset -= now.weekday()
+            if operator in ('+', '-'):
+                if weekday_offset > 0 and operator == '-':
+                    weekday_offset -= 7
+                elif weekday_offset < 0 and operator == '+':
+                    weekday_offset += 7
+            elif isinstance(now, datetime):
+                now += TRUNCATE_TODAY
+            now += timedelta(weekday_offset)
+            continue
+        # Operations on dates
+        try:
+            field_name = _SHORT_DATE_UNIT[term[-1]]
+            number = int(term[1:-1])
+            if operator == '+':
+                pass
+            elif operator == '-':
+                number = -number
+            else:
+                field_name = field_name.removesuffix('s')
+                if isinstance(now, datetime):
+                    if field_name == 'second':
+                        now = now.replace(microsecond=0)
+                    elif field_name == 'minute':
+                        now = now.replace(microsecond=0, second=0)
+                    elif field_name == 'hour':
+                        now = now.replace(microsecond=0, second=0, minute=0)
+                    else:
+                        now += TRUNCATE_TODAY
+                # note: 'week' is not supported
+            now += relativedelta(**{field_name: number})
```

Comment:
You may generalize the logic of `TRUNCATE_TODAY` to simplify code:
```suggestion
            unit = _SHORT_DATE_UNIT[term[-1]]
            if operator in ('+', '-'):
                number = int(term[:-1])  # positive or negative
            else:
                unit = unit.removesuffix('s')
                number = int(term[1:-1])
                if isinstance(now, datetime):
                    now += TRUNCATE_UNIT[unit]
                # note: 'week' is not supported
            now += relativedelta(**{unit: number})
```
where `TRUNCATE_UNIT` is defined declaratively as
```py
TRUNCATE_UNIT = {
    'day': TRUNCATE_TODAY,
    'month': TRUNCATE_TODAY,
    'year': TRUNCATE_TODAY,
    'week': TRUNCATE_TODAY,
    'hour': relativedelta(microsecond=0, second=0, minute=0),
    'minute': relativedelta(microsecond=0, second=0),
    'second': relativedelta(microsecond=0),
}
```